### PR TITLE
ci: release strategy refactored

### PR DESCRIPTION
**Pull request by semoal**
### Done
- [X] Added a `main-suite` workflow:
	- That is responsible of running all the linting, testing, building, and codecoverage on every pull request
	- Also is responsible when running the action manually of deploying the packages to "next" | "latest", (this is defined where the action is run)
	- Also has a matrix-strategy where all the suite is run for node 10,12 and 14.

- [X] Added `next-sync` workflow:
	- Is responsible of syncing `next` branch with `main`, when we push something to `main` automatically will create a pr with all the changes pointing to `next` 👍🏻 

- [X] Added `deploy-main-docs` workflow:
	- Is responsible of deploying automatically the documentation of `main` branch only when someone pushes any change in `/docs/**/*` folder, super cool.
	- `It can be easily changed to deploy to differnt subdomain of vercel if the branch is next`, I'm not an expert of vercel cli or configuration so that part is pretty dark to me, but I can help configuring the dynamism of the Github Action

- [x] Add `pr-lint` pr-lint action will check that the title of the pull request follow our standards, in that way, we only we'll need to merge that pr, and we'll be done. Example of use https://github.com/puppeteer/puppeteer/pull/6540
- [X] Introduced logic of [skip ci] for every workflow 
- [x] Decide if we want to do the full main-suite to every node version (for example we can skip building?, or integration?, dunno) `Right now, I put build, integration and release only for node 14 (lts)`
- [X] Create a `lingui-bot` account to do the commit deployments where's the changelog, package.json changes.. https://github.com/lingui-bot

### Todos

- [ ] Some pending environment variables, i have to check.
We have to define this secrets:
```yml
GH_BOT_PWD: "",
NPM_TOKEN: "",
NOW_TOKEN: ""
```
@tricoder42 i send you the gh_bot_pwd via email 💯 

- [ ] A careful read to check everything is fine.

### CODE REVIEW

File: .github/workflows/main-suite.yml
Comment (tricoder42) [inline .github/workflows/main-suite.yml:54]: Is worth it to `lint` files on all Node.js versions?

File: .github/workflows/next-sync.yml
Comment (tricoder42) [inline .github/workflows/next-sync.yml:11]: Shouldn't this actions run anytime there's a new commit in `main`? No matter if it's marked as `[skip ci]` or not.

File: .github/workflows/deploy-main-docs.yml
Comment (tricoder42) [inline .github/workflows/deploy-main-docs.yml:43]: Is it necessary to install Node when we're building documentation with Python? Even the `npx vercel` command is going to fetch it's own dependencies, isn't it?

```suggestion
```

File: .github/workflows/deploy-main-docs.yml
Comment (tricoder42) [inline .github/workflows/deploy-main-docs.yml:1]: Do I undestand it correctly that the whole point of building the documentation in GitHub actions rather than in Vercel is that we can control *when* the documentation is build? I.e. only when files inside `docs` have changed and there isn't `[skip ci]` in commit message?

File: .github/workflows/release.yml
Comment (tricoder42) [inline .github/workflows/release.yml:27]: What about using the latest LTS here? Can't we specify the version as `latest` or sth?

```suggestion
          node-version: 14
```

File: .github/workflows/release.yml
Comment (tricoder42) [inline .github/workflows/release.yml:62]: What does this action do? Can't we use a `GH_TOKEN` for authentication?

**Review by tricoder42 — COMMENTED** [review]
This is great! I'm really looking forward seeing this machinery up and runnin.

I've added @lingui-bot to @lingui organization and it's `GH_TOKEN` to secrets. `NPM_TOKEN` is still from my personal account, but I guess it's fine.

Just left few comments to clarify some things, but overall — amazing job! Thank you for tackling this 👊

**Comment by tricoder42** [comment]
> Decide if we want to do the full main-suite to every node version (for example we can skip building?, or integration?, dunno) Right now, I put build, integration and release only for node 14 (lts)

I would run tests on all versions and relese only on the latest

> It can be easily changed to deploy to differnt subdomain of vercel if the branch is next, I'm not an expert of vercel cli or configuration so that part is pretty dark to me, but I can help configuring the dynamism of the Github Action

It seems we should add [`--prod`](https://vercel.com/docs/cli#commands/overview/unique-options/prod) when deploying to prod domain

File: .github/workflows/main-suite.yml
Comment (semoal) [reply .github/workflows/main-suite.yml:54]: Absolutely not, I'll lint only for LTS

File: .github/workflows/next-sync.yml
Comment (semoal) [reply .github/workflows/next-sync.yml:11]: Hmm, I'm not sure because the [skip ci] commits will include package.json changes, main changelog, something that we must discard when merging to next. So it's not relevant the main changes into next.

File: .github/workflows/deploy-main-docs.yml
Comment (semoal) [reply .github/workflows/deploy-main-docs.yml:43]: Totally, has sense

File: .github/workflows/deploy-main-docs.yml
Comment (semoal) [reply .github/workflows/deploy-main-docs.yml:1]: Yes, i think is more interesting to control in the CI when the docs are deployed. Now Vercel is deploying the docs always also when the docs doesn't change, so probabbly this strategy it's better.

Comment (semoal) [reply .github/workflows/deploy-main-docs.yml:1]: Vercel should continue creating the temporal domains for pull requests, but for main/next has more sense to me, to control it on the ci. But as you want, both options are valid.

File: .github/workflows/release.yml
Comment (semoal) [reply .github/workflows/release.yml:27]: Actually not, their docs says:

```
It will first check the local cache for a semver match. The hosted images have been updated with the latest of each LTS from v8, v10, v12, and v14.
```

File: .github/workflows/release.yml
Comment (semoal) [reply .github/workflows/release.yml:62]: This is basically doing

git config --add user.email=linguibot@gmail.com
git config --add user.pwd=github password account

We can't use lingui bot gh token, because will override the internal github token of the ci.

Comment (semoal) [reply .github/workflows/release.yml:62]: By default, if we use the github token, the commit will be like this:


INPUT_EMAIL=${INPUT_EMAIL:-'github-action@users.noreply.github.com'}
INPUT_NAME=${INPUT_NAME:-'GitHub Action'}

File: .github/workflows/next-sync.yml
Comment (tricoder42) [reply .github/workflows/next-sync.yml:11]: Hmm, I thought that `next` should be always the same or ahead of `main`. What happens e.g after we release the new version, the `main` will have version `4.0.0`, but `next` will still have version `3.1.4`. Wouldn't it cause troubles down the line?

Ah, I think I got it. Since we have a branching model similar to [git flow](https://nvie.com/posts/a-successful-git-branching-model/), I recall that by default the main branch isn't merged back to develop (similar to what  this action would be). However, it works only because hotfixes (which are direclty merged to `main`) are merged back to `develop`.

I usually use the `git-flow-avh` extension, which automatically merges `main` to `develop` anytime there's a new change. That way I can run `git describe` and see how far ahead of `main` I'm.

Comment (semoal) [reply .github/workflows/next-sync.yml:11]: Yes, I usually worked with git flow strategy that's why I introduced that.

File: .github/workflows/release.yml
Comment (tricoder42) [reply .github/workflows/release.yml:62]: Alright, I've just added `GH_BOT_PWD` to secrets. I just feel very unsecure by adding this password anywhere :/

File: package.json
Comment (tricoder42) [inline package.json:24]: Does Lerna uses semantic-release internally or does it handle similar process on it's own (e.g. generating Changelogs, etc)?

**Review by tricoder42 — APPROVED** [review]
Alright, let's push it and see how it works 👍

File: package.json
Comment (semoal) [reply package.json:24]: Yes does it internally, much better because works out of the box with monorepos.

File: .github/workflows/release.yml
Comment (semoal) [reply .github/workflows/release.yml:62]: If you want we can do it manually, doing the config.set without any external action. It's pretty simple

Comment (tricoder42) [reply .github/workflows/release.yml:62]: That's fine. Let's ship it :) Can't wait to see it in action! 👍

File: .github/workflows/next-sync.yml
Comment (tricoder42) [inline .github/workflows/next-sync.yml:28]: Hey @semoal, sorry to bother you, but could you please explain to me again what's the benefit of using password over a token?

I've checked the `gha-git-credentials` repository and they use it in their example:

```yml
    # publish to a branch in current repo using GITHUB_TOKEN and other default settings
    - uses: oleksiyrudenko/gha-git-credentials@v1
      with:
        token: '${{ secrets.GITHUB_TOKEN }}'
```

Comment (semoal) [reply .github/workflows/next-sync.yml:28]: oh, didn't see that, i only saw the first case that was the case of remote repo. Sorry mate was my fault

Comment (semoal) [reply .github/workflows/next-sync.yml:28]: I change it now

Comment (semoal) [reply .github/workflows/next-sync.yml:28]: I have one question about that because if we use the GITHUB_TOKEN, will do the login correctly to the linguibot account?

Comment (semoal) [reply .github/workflows/next-sync.yml]: I changed it to github token, looks if we are in the same repo/org works perfectly with GITHUB_TOKEN and we'll override the committer.

Comment (semoal) [reply .github/workflows/next-sync.yml]: Nevermind my last sentence, that's wrong I've tried on my lingui fork and the pull request is created by the github action bot account, it's not using the lingui git account.

Comment (semoal) [reply .github/workflows/next-sync.yml]: we can create a personal access token on lingui bot account, to use that instead of the password I guess. what do you think?

Comment (tricoder42) [reply .github/workflows/next-sync.yml]: I've already created a personal access token, but I've add it as `GH_TOKEN` and not `GITHUB_TOKEN`. I've just checked that I can't add `GITHUB_TOKEN`, I guess it's reserved for GitHub Actions account.

Comment (semoal) [reply .github/workflows/next-sync.yml]: Perfect so, I'll change the name asap

Comment (semoal) [reply .github/workflows/next-sync.yml]: Done Tom

**Review by tricoder42 — APPROVED** [review]
Looks good, let's try it 👍

**Comment by tricoder42** [comment]
Just to confirm: What happens after I merge it? Do I need to setup anything manually? I should probably just disable vercel and circleci automatic builds, right?

**Comment by semoal** [comment]
> Just to confirm: What happens after I merge it? Do I need to setup anything manually? I should probably just disable vercel and circleci automatic builds, right?

How main is the default branch automstically will find the new workflows and Will be done. Yes we can disable that

**Comment by tricoder42** [comment]
Cool, let's 🚢  it! Good work!

**Comment by tricoder42** [comment]
The `next-sync` action failed. Not sure why. I've created the `next` branch, but it's still failing. I'll take a look later today.

**Pull request merged**
